features/storage/kvstore/filesystemstore/FileSystemStore.cpp

```diff
@@ -0,0 +1,629 @@
+        status = MBED_ERROR_INVALID_ARGUMENT;
+        goto exit_point;
+    }
+
+    /* If File Exists and is Valid, then check its Write Once Flag to verify its disabled before setting */
+    /* If File exists and is not valid, or is Valid and not Write-Onced then erase it */
+    status = _verify_key_file(key, &key_metadata, &kv_file);
+
+    if (status == MBED_ERROR_INVALID_ARGUMENT) {
+        tr_error("File Verification failed, status: %d", status);
+        goto exit_point;
+    }
+
+    if (status == MBED_SUCCESS) {
+        tr_info("File: %s, Exists. Verifying Write Once Disabled before setting new value", _full_path_key);
+        if (key_metadata.user_flags & KVStore::WRITE_ONCE_FLAG) {
+            kv_file.close();
+            status = MBED_ERROR_WRITE_PROTECTED;
+            goto exit_point;
+        }
+    }
+
+    /* For Success (not write_once) and for corrupted data close file before recreating it as a new file */
+    if (status != MBED_ERROR_ITEM_NOT_FOUND) {
+        kv_file.close();
+    }
+
+    if ((status = kv_file.open(_fs, _full_path_key, O_WRONLY | O_CREAT | O_TRUNC)) != MBED_SUCCESS) {
+        tr_info("set_start failed to open: %s, for writing, err: %d", _full_path_key, status);
+        status = MBED_ERROR_FAILED_OPERATION ;
+        goto exit_point;
+    }
+    _cur_inc_data_size = 0;
+
+    set_handle = new inc_set_handle_t;
+    set_handle->create_flags = create_flags;
+    set_handle->data_size = final_data_size;
+    key_len = strlen(key);
+    set_handle->key = string_ndup(key, key_len);
+    *handle = (set_handle_t)set_handle;
+    _cur_inc_set_handle = *handle;
+
+    key_metadata.magic = FSST_MAGIC;
+    key_metadata.metadata_size = sizeof(key_metadata_t);
+    key_metadata.revision = FSST_REVISION;
+    key_metadata.user_flags = create_flags;
+    kv_file.write(&key_metadata, sizeof(key_metadata_t));
+    kv_file.close();
+exit_point:
+    if (status != MBED_SUCCESS) {
+        _mutex.unlock();
+    }
+    return status;
+}
+
+int FileSystemStore::set_add_data(set_handle_t handle, const void *value_data, size_t data_size)
+{
+    int status = MBED_SUCCESS;
+    size_t added_data = 0;
+    inc_set_handle_t *set_handle = (inc_set_handle_t *)handle;
+    File kv_file;
+
+    if (((value_data == NULL) && (data_size > 0)) || (handle == NULL) || (handle != _cur_inc_set_handle)) {
+        status = MBED_ERROR_INVALID_ARGUMENT;
+        goto exit_point;
+    }
+
+    // Single key incrementally edited, can be edited from multiple threads - lock to protect
+    _inc_data_add_mutex.lock();
+    if ((_cur_inc_data_size + data_size) > set_handle->data_size) {
+        tr_warning("Added Data(%d) will exceed set_start final size(%d) - not adding data to file: %s",
+                   _cur_inc_data_size + data_size, set_handle->data_size, _full_path_key);
+        status = MBED_ERROR_INVALID_SIZE;
+        goto exit_point;
+    }
+
+    if (0 != kv_file.open(_fs, _full_path_key, O_WRONLY | O_APPEND)) {
```

Comment:
It looks like we should be able to open the file in "set_start" and close in "set_finalize". Do we need to open the file every "set_add_data"?

Reply:
Good point. Will add to the list of future optimizations.

Reply:
Future is now. Done.